### PR TITLE
KAFKA-4750: Bypass null value and treat it as deletes

### PR DESCRIPTION
**Pull request by guozhangwang**
Here is the new rule for handling nulls:
* in the interface store, put(key, null) are handled normally and value serde applied to null.
* in the inner most store, null bytes after serialization will always be treated as deletes.
* in the interface store, if null bytes get returned in get(key), serde will be avoided and null object will be returned.

More changes:
* Update javadocs, add unit tests accordingly; augment MockContext to set serdes for the newly added tests.
* Fixed a discovered bug which is exposed by the newly added tests.
* Use the new API to remove all old APIs in the existing state store tests.
* Remove SerializedKeyValueIterator since it is not used any more.

This is originally contributed by @evis.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @dguy @evis This is to complete what's left in https://github.com/apache/kafka/pull/3430. I'm still fighting with the new unit tests, but would like to have you guys take a look at the semantic changes (reflected on the java docs) and see whether they make sense as we discussed on the JIRA ticket.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/261/

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 1]
```diff
@@ -91,10 +91,14 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final Intern
...
-                final V oldValue = sendOldValues ? serdes.valueFrom(underlying.get(entry.key())) : null;
+                V oldValue = null;
+                if (sendOldValues) {
+                    final byte[] oldBytesValue = underlying.get(entry.key());
+                    oldValue = oldBytesValue == null ? null : serdes.valueFrom(oldBytesValue);
+                }
                 underlying.put(entry.key(), entry.newValue());
                 flushListener.apply(serdes.keyFrom(entry.key().get()),
-                                    serdes.valueFrom(entry.newValue()),
+                                    entry.newValue() == null ? null : serdes.valueFrom(entry.newValue()),
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: Wondering if we need this check. It's of course safer, but the contract should be, that the Serde return `null` on `null` ? Thus, it would be a user error? (Same for all other places where the check is added.) Not saying we should have have it, just want to double check about the reasoning.

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java [hunk 2]
```diff
@@ -157,12 +156,16 @@ private void prepareConfigs() {
...
     private static final int TIMEOUT_MULTIPLIER = 5;
 
-    private final TestCondition consumerGroupInactiveCondition = new TestCondition() {
+    private class ConsumerGroupInactiveCondition implements TestCondition {
         @Override
         public boolean conditionMet() {
-            return adminClient.describeConsumerGroup(testId + "-result-consumer", 0).consumers().get().isEmpty();
+            try {
+                return adminClient.describeConsumerGroup(appID, 0).consumers().get().isEmpty();
+            } catch (final TimeoutException e) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: This fix seems unrelated to the null-delete semantics thing -- I think we should split it into two PRs.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java [hunk 3]
```diff
@@ -58,6 +58,27 @@ public void after() {
...
+        store.put(1, null);
+        final KeyValueIterator<Integer, String> it = store.all();
+
+        while (it.hasNext()) {
+            final KeyValue<Integer, String> next = it.next();
+            // should not return deleted records in iterator
+            if (!next.key.equals(2)) {
+                fail("Got deleted key from iterator");
+            }
+        }
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: The iterator should return exactly one record. This, we should add an `Assert.assertFalse(it.hasNext());` after the `if`

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: Yes, this is not intentional; do not know how it gets in, will remove.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: Yes I used the new util just introduced in another PR.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: We actually talked about this on the JIRA and I thought you're preferring this option :) actually, that may not be true. What we proposed is:

```
1) Clarify in javadoc that null value in put(key, value) indicates deletion; if it is "null" object by-pass the serde and send "null" bytes directly into inner functions and vice verse for deserialization; do not enforce user customized serdes how to handle null values since we are not going to call them with null values any more.

2) Do NOT clarify in java doc that null value in put(key, value) indicates deletion; implement all delete(key) call directly throughout all the layers of stores instead of calling put(key, null); recommend user customized serdes to handle null values themselves.
```

I'm following option 1) here, and I think what you suggested here is a bit different with both of them:

```
3) Clarify that any user customized serde must map null to null; if a user failed to do so it is her problem.
```

I personally am inclined to 1) over 3).

**Comment by guozhangwang** [comment]
Added unit tests. Ready for review now.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java [hunk 4]
```diff
@@ -31,13 +31,6 @@ public MemoryNavigableLRUCache(String name, final int maxCacheSize, Serde<K> key
         super(name, maxCacheSize, keySerde, valueSerde);
     }
 
-    @Override
-    public MemoryNavigableLRUCache<K, V> whenEldestRemoved(EldestEntryRemovalListener<K, V> listener) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java:35]: This is intentional since the inherited class as the same function, so no need to override.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueBytesStore.java [hunk 5]
```diff
@@ -45,6 +45,17 @@ public void init(final ProcessorContext context, final StateStore root) {
                 ProcessorStateManager.storeChangelogTopic(
                     context.applicationId(),
                     inner.name())));
+
+        // if the inner store is an LRU cache, add the eviction listener to log removed record
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueBytesStore.java:49]: This is intentional which is a bug fix that gets exposed when adding the unit tests.

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueBytesStore.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueBytesStore.java:49]: In KIP-182 refactoring, for in-memory-lru store we have changed the layering such that this class will embed the `MemoryNavigableLRUCache ` as its inner directly, and the old class is deprecated; however we forgot to add the same logic to register the listener, and hence flushing will not send tombstones to downstream.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/283/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/285/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/284/

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/292/

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java [hunk 6]
```diff
@@ -33,7 +33,7 @@
      * Update the value associated with this key
      *
      * @param key The key to associate the value to
-     * @param value The value, it can be null.
+     * @param value The value to update; if null it is equal as {@link #delete(K)}
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: nit: as -> to

File: streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java [hunk 7]
```diff
@@ -30,15 +30,15 @@
      * Put a key-value pair with the current wall-clock time as the timestamp
      * into the corresponding window
      * @param key The key to associate the value to
-     * @param value The value, it can be null
+     * @param value The value to update; if null it is equal as deleting the value from the corresponding window
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java]: nit: as -> to

File: streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java [hunk 8]
```diff
@@ -30,15 +30,15 @@
...
      * @throws NullPointerException If null is used for key.
      */
     void put(K key, V value);
 
     /**
      * Put a key-value pair with the given timestamp into the corresponding window
      * @param key The key to associate the value to
-     * @param value The value, it can be null
+     * @param value The value; if null it is equal as deleting the value from the corresponding window
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java]: same as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: My inclination is to go with option 2 as we discussed in the Jira, but if it's between option 1 and 3 presented here then I'll take 1 over 3 as well

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStoreTest.java [hunk 9]
```diff
@@ -36,27 +35,11 @@
...
-
-        if (useContextSerdes) {
-            keySerde = (Serde<K>) context.keySerde();
-            valueSerde = (Serde<V>) context.valueSerde();
-        } else {
-            keySerde = Serdes.serdeFrom(keyClass);
-            valueSerde = Serdes.serdeFrom(valueClass);
-        }
-
+    protected <K, V> KeyValueStore<K, V> createKeyValueStore(ProcessorContext context) {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStoreTest.java]: This is intentional: we get rid of the other parameters since they are not useful any more.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: I was not sure if we did agree on anything... Also just reread the discussion. I don't think that option (2) works, because we inherit delete semantics from the broker compaction and thus, cannot avoid it.

The question is really if you put the burden on the user to implement a correct Serde (that must map `null` to `null`), or if we do the check within the library (safer, but also some runtime overhead...).

If we go with (1) should we add checks that serializer does never return `null` ?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStoreTest.java (hunk 9 above)
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStoreTest.java]: nit: make parameter `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreTest.java [hunk 10]
```diff
@@ -30,22 +32,15 @@
 
     @SuppressWarnings("unchecked")
     @Override
-    protected <K, V> KeyValueStore<K, V> createKeyValueStore(
-            ProcessorContext context,
-            Class<K> keyClass,
-            Class<V> valueClass,
-            boolean useContextSerdes) {
+    protected <K, V> KeyValueStore<K, V> createKeyValueStore(ProcessorContext context) {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryLRUCacheStoreTest.java [hunk 11]
```diff
@@ -36,22 +38,17 @@
...
-            Class<V> valueClass,
-            boolean useContextSerdes) {
-
-        StateStoreSupplier supplier;
-        if (useContextSerdes) {
-            supplier = Stores.create("my-store").withKeys(context.keySerde()).withValues(context.valueSerde()).inMemory().maxEntries(10).build();
-        } else {
-            supplier = Stores.create("my-store").withKeys(keyClass).withValues(valueClass).inMemory().maxEntries(10).build();
-        }
+    protected <K, V> KeyValueStore<K, V> createKeyValueStore(ProcessorContext context) {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryLRUCacheStoreTest.java]: as above

File: streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java [hunk 12]
```diff
@@ -121,6 +121,14 @@ public RecordCollector recordCollector() {
         return recordCollector;
     }
 
+    public void setKeySerde(Serde<?> keySerde) {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java]: nit: parameter final here and below

**Comment by bbejeck** [comment]
Left some comments

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: About 2), yes, we dropped this option exactly because kafka side log compaction does treat `null` value specially anyways, and hence we decided to go with 1) (at least that was my understanding..) I'm only copying the content for context in the previous reply, and assuming that we have selected 1) over 2).

And now about 1) v.s. 3), like I said I think 3) is less efficient since it requires everyone to implement `null` to `null` mapping which is too much, so I'm still inclined to 1).

As for implementation, even if user mapped `not-null` to `null` in ser, the most inner stores (memory map and rocks) would still treat `null` bytes specially as `delete`; as for deser, if the user maps `not-null` bytes to `null` objects, then we would just return that in the api call and that should be fine since it is user's call to interpret `null` object as "not exist" or "a dummy value" then.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: Yes. So if users have buggy ser code and return `null` for non-`null` input, they effectively delete data accidentally. Should we add a note to Serde/Serializer JavaDocs (web docs)? As an alternative, we should even check if ser returns `null` for non-`null` input and throw error to notify users about their buggy ser code?

I am fine with going with (1) -- might be better to protect users from potentially buggy serde code. Just wanted to clarify our decision.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 13]
```diff
@@ -212,15 +216,15 @@ public void put(final Bytes key, final byte[] value) {
...
         try {
+            // for null bytes, we still put it into cache indicating tombstones
             putInternal(key, value);
         } finally {
             lock.writeLock().unlock();
         }
     }
 
-    private void putInternal(final Bytes rawKey, final byte[] value) {
-        Objects.requireNonNull(rawKey, "key cannot be null");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:222]: We only need to check this in public functions, internal ones do not need to check again.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 14]
```diff
@@ -254,19 +261,22 @@ public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
...
-            cache.delete(cacheName, key);
-            underlying.delete(key);
-            return v;
+            return deleteInternal(key);
         } finally {
             lock.writeLock().unlock();
         }
     }
 
+    private byte[] deleteInternal(final Bytes key) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:274]: This is also intentional but I'd like to have at least a few more careful checks on: I think our current handling of delete, which is to remove from cache and remove from store immediately is not correct, since by doing this, the deletion will not be forwarded to downstream processors (even if the store is flushed later). So I changed it to just putting a null byte arrays into the cache, which will also be picked over the underlying store in the merge-sort iterators.

But I want to make sure that there are no other corner cases that may be affected by this change. @dguy @mjsax @bbejeck ?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/303/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/305/

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java [hunk 15]
```diff
@@ -105,7 +105,10 @@ public synchronized V get(K key) {
 
     @Override
     public synchronized void put(K key, V value) {
-        this.map.put(key, value);
+        if (value == null)
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: nit: java formatting?

**Review by dguy — APPROVED** [review]
just one nit, otherwise LGTM

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 14 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:274]: I believe this to be correct as well, as all the underlying stores handle the put of a `null` as a delete and it seems compatible with the new change to not have `all()` include null values.

**Comment by bbejeck** [comment]
Thanks, @guozhangwang LGTM

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java [hunk 16]
```diff
@@ -33,7 +33,7 @@
      * Update the value associated with this key
      *
      * @param key The key to associate the value to
-     * @param value The value, it can be null.
+     * @param value The value to update; can be null
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: Might this be miss-leading? `can be null` does not indicate that `null` has special delete semantics... (also on other places)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 14 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:274]: I am not 100% familiar with the details here, but it sounds reasonable. If you are correct and our old code is incorrect, we should write a test that fails with the old code but passed with the new code.

Or is this exposed already by a newly added test as mentioned here https://github.com/apache/kafka/pull/4508/files#r165542967

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java [hunk 17]
```diff
@@ -136,13 +136,17 @@ public boolean isOpen() {
...
         Objects.requireNonNull(key);
+
         return this.map.get(key);
     }
 
     @Override
     public synchronized void put(K key, V value) {
         Objects.requireNonNull(key);
-        this.map.put(key, value);
+        if (value == null)
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java]: nit: add `{ }` to both blocks

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java [hunk 18]
```diff
@@ -105,7 +105,10 @@ public synchronized V get(K key) {
 
     @Override
     public synchronized void put(K key, V value) {
-        this.map.put(key, value);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: add `final` to parameters for further code cleanup (maybe others parameters in this class, too)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java [hunk 19]
```diff
@@ -136,13 +136,17 @@ public boolean isOpen() {
     @Override
     public synchronized V get(K key) {
         Objects.requireNonNull(key);
+
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java]: add `final` to parameters for further code cleanup (maybe others parameters in this class, too)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java (hunk 15 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: ACK.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:274]: The reason I think we did not capture this issue earlier, is that `tuple forward in flushing` is only used for DSL, and in our current DSL's KTableXX operators we never explicitly trigger `delete` and hence that's why we never realize this issue.

But even without the scenario that could expose this bug, not using write-through caching for delete still have a performance benefits of write-back caching.

File: streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java [hunk 20]
```diff
@@ -121,6 +121,14 @@ public RecordCollector recordCollector() {
         return recordCollector;
     }
 
+    public void setKeySerde(final Serde<?> keySerde) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java:124]: Why do we add this? Seems to be unused?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java [hunk 21]
```diff
@@ -67,6 +69,34 @@ public void after() {
         return result;
     }
 
+    @Test
+    public void testDelete() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: nit: `shouldDeleteIfSerializedValueIsNull`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java [hunk 22]
```diff
@@ -67,6 +69,34 @@ public void after() {
...
+            }
+        };
+
+        context.setValueSerde(Serdes.serdeFrom(serializer, new StringDeserializer()));
+        store = createKeyValueStore(driver.context());
+
+        store.put(0, "zero");
+        store.put(1, "one");
+        store.put(2, "two");
+        store.delete(0);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: Should we add a test for `delete` to make sure it does not call the serializer?

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: `null` object may not necessarily be serialized to `null` bytes, and hence I did not mention it indicates deletes. But I can further mention that if the serialized bytes is `null` it indicates deletes.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java (hunk 18 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: ack.

File: streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java (hunk 20 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java:124]: It is added to be symmetry with `setValueSerde`.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java (hunk 21 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: It actually test more than that: both direct `delete` and `put with null bytes`.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java (hunk 22 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: Not sure I understand.. do you mean the serializer for the key is not called (which is not the case)? Or the serializer of the value (we do not have value parameter)?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: I referred to the value serializer.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java (hunk 21 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: Split into two test? Or rename accordingly.`testDelete` is not a good name IMHO.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/359/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/360/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/350/

**Review by mjsax — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
Cherry-picked to 1.1.0 as well.
